File: examples.ncs/getting-started/developing-with-ncs/1-cdb-py/packages/cdb/python/plaincdbsub/plaincdbsub.py
# -*- mode: python; python-indent: 4 -*-

import _ncs
import _ncs.cdb as cdb
import _ncs.maapi as maapi
from ncs_pyvm import NcsPyVM

import _namespaces.test_ns as ns

import socket
import threading

_schemas_loaded = False

# This low level cdb subscriber subscribes to changes under the path
# /devices/device{ex0}/config
# Whenever a change occurs there, the code iterates through the
# change and prints the values. Thus to trigger this subscription code
# go into the ncs_cli and commit any change under the subscription
# path. For example:

# ncs_cli -u admin
# admin connected from 127.0.0.1 using console on iron.local
# admin@iron> configure
# dmin@iron% set devices device ex0 config sys syslog server 4.5.6.7 enabled
# [ok][2012-07-05 12:57:59]

# [edit]
# admin@iron% commit
# Commit complete.

# will trigger the subscription code, the code logs and the data will end up
# in ./logs/ncs-python-vm-cdb.log (relative to where the ncs daemon executes)

# The code runs in an 'application' component, it implements
# the ApplicationComponent interface, this includes the run() method
# so the code will run in a Thread.


class Subscriber(object):
    def __init__(self, prio, path, debug):
        self.sock = socket.socket()
        self.path = path
        self.debug = debug
        self.prio = prio
        global _schemas_loaded

        if _schemas_loaded is False:
            ms = socket.socket()
            maapi.connect(sock=ms, ip='127.0.0.1',
                          port=_ncs.NCS_PORT)
            maapi.load_schemas(ms)
            ms.close()
            _schemas_loaded = True

        cdb.connect(self.sock, type=cdb.DATA_SOCKET, ip='127.0.0.1',
                    port=_ncs.NCS_PORT, path=self.path)
        self.subid = cdb.subscribe(self.sock, prio=self.prio, path=self.path,
                                   nspace=ns.ns.hash)
        cdb.subscribe_done(self.sock)
        self.debug("Subscription {0}, subscribed to {1}".format(self.subid,
                                                                self.path))

    def wait(self):
        cdb.read_subscription_socket(self.sock)

    def ack(self):
        cdb.sync_subscription_socket(self.sock, cdb.DONE_PRIORITY)

    def diff_iter(self):
        def iterator(kp, op, oldv, newv, state):
            self.debug(
                "diffIterate: kp= {0}, OP= {1}, old_value={2}, new_value={3}"
                .format(str(kp), op, str(oldv), str(newv)))
            return _ncs.ITER_RECURSE
        cdb.diff_iterate(self.sock, self.subid, iterator, 0, None)
    
    def diff_iter_loop(self):
        while True:
            self.wait()
            self.diff_iter()
            self.ack()

    def close(self):
        cdb.end_session(self.sock)
        cdb.close(self.sock)


# ---------------------------------------------
# COMPONENT THREAD THAT WILL BE STARTED BY NCS.
# ---------------------------------------------

class Plaincdbsub(object):
    def __init__(self, *args, **kwds):
        # Setup the NCS object, containing mechanisms
        # for communicating between NCS and this User code.
        self._ncs = NcsPyVM(*args, **kwds)
        self.debug('--- Service INIT OBJECT')
        # Register our 'finish' callback
        self._finish_cb = lambda: self.finish()

        self.sub = Subscriber(prio=100, path="/devices/device{ex0}/config",
                              debug=self.debug)

        self._ncs.reg_finish(self._finish_cb)

        self._stopevent = threading.Event()

    # This method is supposed to start the User application
    def run(self):
        self.debug('Running diff iter loop')
        self.sub.diff_iter_loop()

    # Just a convenient logging function
    def debug(self, line):
        self._ncs.debug(line)

    # Callback that will be invoked by NCS when the system is shutdown.
    # Make sure to shutdown the User code, including any User created threads.
    def finish(self):
        self.debug(' PlainSub in finish () =>\n')
        self.sub.close()
        self._stopevent.set()
        self.debug(' PlainSub in finish () => ok\n')

File: examples.ncs/getting-started/developing-with-ncs/1-cdb/packages/cdb/src/java/src/com/example/cdb/PlainCdbSub.java
package com.example.cdb;

import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.LogManager;

import com.tailf.cdb.Cdb;
import com.tailf.cdb.CdbDiffIterate;
import com.tailf.cdb.CdbSubscription;
import com.tailf.cdb.CdbSubscriptionSyncType;
import com.tailf.conf.Conf;
import com.tailf.conf.ConfException;
import com.tailf.conf.ConfObject;
import com.tailf.conf.DiffIterateOperFlag;
import com.tailf.conf.DiffIterateResultFlag;
import com.tailf.conf.ErrorCode;
import com.tailf.ncs.ApplicationComponent;
import com.tailf.ncs.ResourceManager;
import com.tailf.ncs.annotations.Resource;
import com.tailf.ncs.annotations.ResourceType;
import com.tailf.ncs.annotations.Scope;
import com.tailf.ncs.ns.Ncs;


// This low level cdb subscriber subscribes to changes under the path
// /devices/device{ex0}/config
// Whenever a change occurs there, the code iterates through the
// change and prints the values. Thus to trigger this subscription code
// go into the ncs_cli and commit any change under the subscription
// path. For example:

// # ncs_cli -u admin
// admin connected from 127.0.0.1 using console on iron.local
// admin@iron> configure
// dmin@iron% set devices device ex0 config sys syslog server 4.5.6.7 enabled
// [ok][2012-07-05 12:57:59]

// [edit]
// admin@iron% commit
// Commit complete.


// will trigger the subscription code, the code logs and the data will end up
// in ./logs/ncs-java-vm.log (relative to where the ncs daemon executes)


// The code runs in an 'application' component, it implements
// the ApplicationComponent interface, this includes the run() method
// so the code will run in a Thread.


public class PlainCdbSub implements ApplicationComponent  {
    private static Logger LOGGER = LogManager.getLogger(PlainCdbSub.class);

    @Resource(type=ResourceType.CDB, scope=Scope.INSTANCE, qualifier="plain")
    private Cdb cdb;

    private CdbSubscription sub;
    private int subId;
    private boolean requestStop = false;

    public PlainCdbSub() {
    }

    public void init() {
        try {
            LOGGER.info(" init cdb subscriber ");
            sub = new CdbSubscription(cdb);
            String str = "/devices/device{ex0}/config";
            subId = sub.subscribe(1, new Ncs(), str);
            sub.subscribeDone();
            LOGGER.info("subscribeDone");
            requestStop = false;
        } catch (Exception e) {
            throw new RuntimeException("FAIL in init", e);
        }
    }

    public void run() {
        try {
            while (!requestStop) {
                try {
                    int[] points = sub.read();
                    sub.diffIterate(subId, new Iter());
                } finally {
                    sub.sync(CdbSubscriptionSyncType.DONE_SOCKET);
                }
            }
        } catch (ConfException e) {
            if (e.getErrorCode() == ErrorCode.ERR_EOF) {
                // Triggered by finish method
                // if we throw further NCS JVM will try to restart
                // the package
                LOGGER.warn(" Socket Closed!");
            } else {
                throw new RuntimeException("FAIL in run", e);
            }
        } catch (Exception e) {
            LOGGER.warn("Exception:" + e.getMessage() );
            throw new RuntimeException("FAIL in run", e);
        } finally {
            requestStop = false;
            LOGGER.warn(" run end ");
        }
    }

    public void finish() {
        requestStop = true;
        LOGGER.warn(" PlainSub in finish () =>");
        try {
            // ResourceManager will close the resource (cdb) used by this
            // instance that triggers ConfException with ErrorCode.ERR_EOF
            // in run method
            ResourceManager.unregisterResources(this);
        } catch (Exception e) {
            throw new RuntimeException("FAIL in finish", e);
        }
        LOGGER.warn(" PlainSub in finish () => ok");
    }

    private class Iter implements CdbDiffIterate  {
        public DiffIterateResultFlag iterate(ConfObject[] kp,
                                             DiffIterateOperFlag op,
                                             ConfObject old_value,
                                             ConfObject new_value,
                                             Object state) {
            try {
                String kpString = Conf.kpToString(kp);
                LOGGER.info("diffIterate: kp= "+kpString+", OP="+op+
                            ", old_value="+old_value+", new_value="+
                            new_value);
                return DiffIterateResultFlag.ITER_RECURSE;
            } catch (Exception e) {
                return DiffIterateResultFlag.ITER_CONTINUE;
            }
        }
    }
}

